### DISCUSSION TITLE
Installation fails - Windows Server 2019, Oqtane 2.1

### DISCUSSION THREAD

**Discussion by MartyNZ** [oqtane/oqtane.framework#1459] [General] [answered]
Hi,
I have repeatedly tried to install Oqtane on a production server (Windows Server 2019) as outlined at ([https://azing.org/oqtane/r/7mk6DNd0])
Once I try to browse to the site (Step 7) I get a HTTP 403.14 error.

Can someone please suggest where the issue is and how to start troubleshooting.

**Comment by sbwalker** [comment, marked as answer]
Perhaps these instructions will help: https://www.oqtane.org/Resources/Blog/PostId/542/installing-oqtane-on-iis

**Reply by MartyNZ** [reply]
Hi Shaun,
I have followed those instructions and get the same result. Step 9 - Navigate to site results in
"This page isn’t working
www.sol.net.nz is currently unable to handle this request.
HTTP ERROR 500"

when browsed to from the internet or,

"HTTP Error 403.14 - Forbidden
The Web server is configured to not list the contents of this directory." when browsed to locally using the hosts file.

**Reply by sbwalker** [reply]
@iJungleboy created a more detailed video which may help: https://2sxc.org/en/blog/post/install-oqtane-2-from-scratch-with-language-packs-and-2sxc-video

**Reply by MartyNZ** [reply]
Yep.
Followed his azing list as well. Only difference to the process is the application pool. And I get exactly the same result...

**Reply by sbwalker** [reply]
So I just tried to access one of the static files in the wwwroot folder and it worked:

http://www.sol.net.nz/loading.gif

You are certain that you have followed EVERY step in the instructions? 

- Within Application Development Features make sure that WebSocket Protocol is enabled ( this is needed for SignalR ) 

- Within Common HTTP Features make sure WebDAV Publishing is disabled ( if it is enabled it will conflict with Blazor ) 

- install .NET Core hosting bundle for .NET 5

**Reply by MartyNZ** [reply]
![Screenshot 2021-06-10 140803](https://user-images.githubusercontent.com/13032309/121453809-dbfea000-c9f5-11eb-9761-e53ebbe94876.jpg)
![Screenshot 2021-06-10 140905](https://user-images.githubusercontent.com/13032309/121453811-dd2fcd00-c9f5-11eb-806d-69b0978ef2c6.jpg)
![Screenshot 2021-06-10 141013](https://user-images.githubusercontent.com/13032309/121453813-ddc86380-c9f5-11eb-939e-7a08a71f7bc1.jpg)
![Screenshot 2021-06-10 141128](https://user-images.githubusercontent.com/13032309/121453814-de60fa00-c9f5-11eb-8486-98bc14720a1b.jpg)
![Screenshot 2021-06-10 141154](https://user-images.githubusercontent.com/13032309/121453815-de60fa00-c9f5-11eb-98bf-521864fcd29e.jpg)

**Reply by MartyNZ** [reply]
If I've missed anything please let me know. I've been over it so many times I may have developed a blind spot...

**Comment by louyo** [comment]
Pardon me for stating what may be obvious:
https://docs.microsoft.com/en-us/troubleshoot/iis/http-403-14-forbidden-webpage

**Reply by MartyNZ** [reply]
You're excused.
a) checked and implemented before I opened this discussion.
b) there is no directive in the installation notes that state this step is required.

Thanks for trying to help though.

**Comment by iJungleboy** [comment]
Hi @MartyNZ 

First of all to start debugging you should set the web.config debugging to true - then look at the /logs folder, that always helped me. 

I ran into a fairly special but important issue when installing on a real server (a problem which doesn't happen on localhost or xyz.oqtane.me installs), which I could easily imagine is your issue. For this I must explain some background. If you're running Razor in Server-Mode (which is probably the most common today) you actually have 3 layers:

1. The **Browser Client** which is basically an empty page and some JS which will then use WebSockets to load html / data from the backend. 
2. The **Server Client** which is an isolated process that does the work which could otherwise run in the browser - this will generate HTML and stuff and do click-handling etc. 
3. The **Server Backend** which contains backend APIs

What I ran into (and I believe many will) is that I used a hosts.ini setting on my PC to test this. So for example I configure `oqtane.test` to point to `192.168.10.10` (that would be the server IP) then this happens:

1. I point my browser to `oqtane.test` - it will resolve `192.168.10.10` and get the **Browser Client** which will then - running on my PC - continue with WebSockets requests to `oqtane.test` - again on `192.168.10.10`.
2. This will start the **Server Client** which is in an isolated process. It will start processing the DLLs at very early on need some data from the backend. The communication between the **Server Client** and the **Server Backend** happens through http, even though it's running on the same machine. 
3. The **Server Client** knows it's communicating on `oqtane.test` so it will open an http-connection to the **Server Backend** on `oqtane.test`. Since this process is running on the server, it doesn't have your hosts.ini setting and will not be able to find this IP. So that would fail. 

This is IMHO one of the biggest issues which people will run into when test-installing it on real servers.

Note that I've just updated the checklists to cover debugging and this special issue https://azing.org/oqtane/r/7mk6DNd0

**Reply by sbwalker** [reply]
@iJungleboy I believe you are referring to the "loopback" or "hairpin" behavior when running Oqtane on Blazor Server - where the "Client" services interact with the "Server" Controllers via HttpClient - however on Blazor Server this is all physically happening on the Server which essentially means the application is calling itself via HttpClient. The framework is architected in this way so that it can support both Blazor WebAssembly and Blazor Server within the same code base - but it seems to cause some extra configuration challenges with the hosts file. This thread also discussed this topic previously: https://github.com/oqtane/oqtane.framework/issues/973

**Reply by MartyNZ** [reply]
Ok, It's now getting a bit weirder...
I have enabled logging and once I browsed to the site **on the hosting server** I was able to see and complete the installation process. However, I can only browse it on the server, browsing from an external source (solutions-online.co.nz) still fails. I've attached the log file which shows a fail when a connection is requested 
[stdout_20210610234626_22300.log](https://github.com/oqtane/oqtane.framework/files/6634966/stdout_20210610234626_22300.log)
from an external source (solutions-online.co.nz)

**Reply by sbwalker** [reply]
@MartyNZ So I see requests to http://solutions-online.co.nz/ and requests to http://sol.net.nz - similar to DNN, Oqtane uses an "alias" concept to map domains to sites. It looks like you initially installed the application using http://solutions-online.co.nz/ - so "solutions-online.co.nz" would have been the initial Alias value which was created. Did you then login to the default site and go to Site Settings and add an alias for "sol.net.nz" ? If you did not then Oqtane would not know how to resolve that address to a specific site.

**Reply by sbwalker** [reply]
Actually I see even more domains in the log:

Now listening on: http://www.solutions-online.co.nz:80/
Now listening on: http://www.sol.net.nz:80/
Now listening on: http://solutions-online.co.nz:80/
Now listening on: http://vnext.solutions-online.co.nz:80/
Now listening on: http://sol.net.nz:80/
Now listening on: http://vnext.sol.net.nz:80/

What are the contents of your Alias table in your database?

**Reply by MartyNZ** [reply]
Other way around.
The site got installed using sol.net.nz (entered into the hosts file and browsed to internally). I then added solutions-online.co.nz as an alias and tried to browse from an external source. I have kept to using sol.net.nz when browsing internally and solutions-online.co.nz when browsing from an external source. Yes, the other aliases have been added to the site but I haven't used any of them at this stage (and yes, all aliases have bindings in IIS).

**Reply by sbwalker** [reply]
@MartyNZ after looking at your log file I can definitely see that when you interact with the app using http://sol.net.nz/ it works fine. However when you try to interact using http://solutions-online.co.nz/ it fails on the very first service invocation to "/api/Installation/isinstalled" and the error is:

"connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. (solutions-online.co.nz:80)"

This would indicate that System.Net.Http.HttpClient failed due to some environment configuration issue. This goes back to what @iJungleboy described earlier in this thread - it looks like your server is either not able to resolve the domain solutions-online.co.nz or make http calls to itself (ie. loopback). I am not sure how to troubleshoot this - can you login to the server and attempt to do a ping or traceroute to http://solutions-online.co.nz?

**Reply by sbwalker** [reply]
Actually @iJungleboy added new instructions to his site:

To Solve this you have these Options:

1. Use XYZ.oqtane.me and test on the PC (as all *.oqtane.me point to localhost)
- Configure the IIS on the server to use such a domain and listen to 127.0.0.1 in addition to your server IP - you don't need a - hosts.ini entry as it's already in the public DNS
- Test it using a browser on the server

2. Use XYZ.oqtane.me and test on your PC
- Configure IIS as mentioned in the first example
- Then configure your local hosts.ini file to have the same domain but point to the server IP
- Now you can test from your local PC 

3. Use a real domain which you added to a real DNS which is used by both your PC and the server

**Reply by MartyNZ** [reply]
Hi @sbwalker,
So I completed the steps as outlined by @iJungleboy. I can now successfully browse to the site both externally and internally (from server and PC) using the domain sol.oqtane.me. Step 3. using solutions-online.co.nz still fails.

FYI: from the server, solutions-online.co.nz successfully pings the servers public ip address and tracert works in a single hop. So all good there.

I've attached the latest log file.
[stdout_20210612004841_13600.log](https://github.com/oqtane/oqtane.framework/files/6641414/stdout_20210612004841_13600.log)

**Reply by sbwalker** [reply]
@MartyNZ did you ever figure out your configuration issue for the solutions-online.co.nz domain?

**Reply by MartyNZ** [reply]
Hi @sbwalker,
Thanks for following up.
At this stage no, I haven't found the cause. I followed the steps you provided from @iJungleboy and the result is I have a publicly visible and functioning Oqtane site at [https://sol.net.nz](https://sol.net.nz). This was achieved by entering the domain in the Servers hosts file.

As this is a production server I'm not willing to make configuration changes until I understand all the consequences. And as I have full access to the server it's an easy workaround. It's not a priority at the moment but I do still want to find the cause, this server hosts DNN sites and will be hosting Oqtane as soon as I'm comfortable with it (usually around v3.2).

The same cannot be said for my Oqtane install on Azure...
New thread coming on that one :-)

Cheers

**Reply by iJungleboy** [reply]
@MartyNZ I have a few guesses :)

To summarize my understanding is that it doesn't work until you add a manual hosts.ini entry on the server. 

Basically this means that when the IIS-Service is doing a DNS request it wouldn't find the right IP until you add your hosts entry. If we go a step back it means that the name - which seems to resolve to that server from the outside - doesn't resolve to the correct target on the inside. 

This is actually kind of common for various reasons usually related to DMZ segregation. From my experience these common DMZ scenarios:

1. Often times a web server doesn't itself need to access the internet - so often times it doesn't actually have a DNS which resolves external names.
2. Often also the DMZ may _be inside a domain_ which has the corporate setup. So many companies have a `www.mycompany.com` and the internal domain is also `mycompany.com`. Usually the external and internal DNS are _not_ the same, because `fileserver1.mycompany.com` should resolve on the inside, but the outside shouldn't know about this. This issue is super-common and has caused us countless support tickets. Often then the external DNS resolves something as expected, but the internal one was never adjusted (or couldn't be adjusted at all - because like in our case `2sic.com` must point to our AD-Controller, but externally `2sic.com` should point to our website. 

good luck 👍

**Reply by MartyNZ** [reply]
@iJungleboy You may be onto something there.
I do know the Network Admin was doing something with DMZ a while back. I'll have a chat with him soon.

Thanks for your feedback.
